File: src/main/java/aurocosh/divinefavor/common/item/spell_bow/ItemSpellBow.java
package aurocosh.divinefavor.common.item.spell_bow;

import aurocosh.divinefavor.DivineFavor;
import aurocosh.divinefavor.common.constants.ConstGuiIDs;
import aurocosh.divinefavor.common.core.DivineFavorCreativeTab;
import aurocosh.divinefavor.common.item.base.ModItem;
import aurocosh.divinefavor.common.item.common.ModItems;
import aurocosh.divinefavor.common.item.spell_bow.capability.ISpellBowHandler;
import aurocosh.divinefavor.common.item.spell_bow.capability.SpellBowProvider;
import aurocosh.divinefavor.common.item.spell_bow.capability.SpellBowStorage;
import aurocosh.divinefavor.common.util.UtilNbt;
import net.minecraft.enchantment.EnchantmentHelper;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.projectile.EntityArrow;
import net.minecraft.init.Enchantments;
import net.minecraft.init.Items;
import net.minecraft.init.SoundEvents;
import net.minecraft.item.EnumAction;
import net.minecraft.item.IItemPropertyGetter;
import net.minecraft.item.ItemArrow;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.stats.StatList;
import net.minecraft.util.*;
import net.minecraft.world.World;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import javax.annotation.Nullable;

import static aurocosh.divinefavor.common.item.spell_bow.capability.SpellBowDataHandler.CAPABILITY_SPELL_BOW;

public class ItemSpellBow extends ModItem {
    public static final String TAG_IS_IN_BOOK_MODE = "IsInBookMode";
    private static String TAG_SHARE = "SpellBowShare";

    public ItemSpellBow()
    {
        super("spell_bow","spell_bow/spell_bow");
        this.maxStackSize = 1;
        setCreativeTab(DivineFavorCreativeTab.INSTANCE);
        this.setMaxDamage(384);

        this.addPropertyOverride(new ResourceLocation("book_mode"), new IItemPropertyGetter()
        {
            @SideOnly(Side.CLIENT)
            public float apply(ItemStack stack, @Nullable World world, @Nullable EntityLivingBase entityLivingBase)
            {
                NBTTagCompound nbt = UtilNbt.getTag(stack);
                return nbt.getBoolean(TAG_IS_IN_BOOK_MODE) ? 1 : 0;
            }
        });
        this.addPropertyOverride(new ResourceLocation("pull"), new IItemPropertyGetter()
        {
            @SideOnly(Side.CLIENT)
            public float apply(ItemStack stack, @Nullable World world, @Nullable EntityLivingBase entityLivingBase)
            {
                if (entityLivingBase == null)
                    return 0.0F;
                else
                    return entityLivingBase.getActiveItemStack().getItem() != ModItems.spell_bow ? 0.0F : (float) (stack.getMaxItemUseDuration() - entityLivingBase.getItemInUseCount()) / 20.0F;
            }
        });
        this.addPropertyOverride(new ResourceLocation("pulling"), new IItemPropertyGetter()
        {
            @SideOnly(Side.CLIENT)
            public float apply(ItemStack stack, @Nullable World world, @Nullable EntityLivingBase entity)
            {
                return entity != null && entity.isHandActive() && entity.getActiveItemStack() == stack ? 1.0F : 0.0F;
            }
        });
    }

    private ItemStack findAmmo(EntityPlayer player)
    {
        if (this.isArrow(player.getHeldItem(EnumHand.OFF_HAND)))
        {
            return player.getHeldItem(EnumHand.OFF_HAND);
        }
        else if (this.isArrow(player.getHeldItem(EnumHand.MAIN_HAND)))
        {
            return player.getHeldItem(EnumHand.MAIN_HAND);
        }
        else
        {
            for (int i = 0; i < player.inventory.getSizeInventory(); ++i)
            {
                ItemStack itemstack = player.inventory.getStackInSlot(i);

                if (this.isArrow(itemstack))
                {
                    return itemstack;
                }
            }

            return ItemStack.EMPTY;
        }
    }

    protected boolean isArrow(ItemStack stack)
    {
        return stack.getItem() instanceof ItemArrow;
    }

    /**
     * Called when the player stops using an Item (stops holding the right mouse button).
     */
    public void onPlayerStoppedUsing(ItemStack stack, World worldIn, EntityLivingBase entityLiving, int timeLeft)
    {
        if(UtilNbt.getTag(stack).getBoolean(TAG_IS_IN_BOOK_MODE))
            return;

        if (entityLiving instanceof EntityPlayer)
        {
            EntityPlayer entityplayer = (EntityPlayer)entityLiving;
            boolean flag = entityplayer.capabilities.isCreativeMode || EnchantmentHelper.getEnchantmentLevel(Enchantments.INFINITY, stack) > 0;
            ItemStack itemstack = this.findAmmo(entityplayer);

            int i = this.getMaxItemUseDuration(stack) - timeLeft;
            i = net.minecraftforge.event.ForgeEventFactory.onArrowLoose(stack, worldIn, entityplayer, i, !itemstack.isEmpty() || flag);
            if (i < 0) return;

            if (!itemstack.isEmpty() || flag)
            {
                if (itemstack.isEmpty())
                {
                    itemstack = new ItemStack(Items.ARROW);
                }

                float f = getArrowVelocity(i);

                if ((double)f >= 0.1D)
                {
                    boolean flag1 = entityplayer.capabilities.isCreativeMode || (itemstack.getItem() instanceof ItemArrow && ((ItemArrow) itemstack.getItem()).isInfinite(itemstack, stack, entityplayer));

                    if (!worldIn.isRemote)
                    {
                        ItemArrow itemarrow = (ItemArrow)(itemstack.getItem() instanceof ItemArrow ? itemstack.getItem() : Items.ARROW);
                        EntityArrow entityarrow = itemarrow.createArrow(worldIn, itemstack, entityplayer);
                        entityarrow.shoot(entityplayer, entityplayer.rotationPitch, entityplayer.rotationYaw, 0.0F, f * 3.0F, 1.0F);

                        if (f == 1.0F)
                        {
                            entityarrow.setIsCritical(true);
                        }

                        int j = EnchantmentHelper.getEnchantmentLevel(Enchantments.POWER, stack);

                        if (j > 0)
                        {
                            entityarrow.setDamage(entityarrow.getDamage() + (double)j * 0.5D + 0.5D);
                        }

                        int k = EnchantmentHelper.getEnchantmentLevel(Enchantments.PUNCH, stack);

                        if (k > 0)
                        {
                            entityarrow.setKnockbackStrength(k);
                        }

                        if (EnchantmentHelper.getEnchantmentLevel(Enchantments.FLAME, stack) > 0)
                        {
                            entityarrow.setFire(100);
                        }

                        stack.damageItem(1, entityplayer);

                        if (flag1 || entityplayer.capabilities.isCreativeMode && (itemstack.getItem() == Items.SPECTRAL_ARROW || itemstack.getItem() == Items.TIPPED_ARROW))
                        {
                            entityarrow.pickupStatus = EntityArrow.PickupStatus.CREATIVE_ONLY;
                        }

                        worldIn.spawnEntity(entityarrow);
                    }

                    worldIn.playSound((EntityPlayer)null, entityplayer.posX, entityplayer.posY, entityplayer.posZ, SoundEvents.ENTITY_ARROW_SHOOT, SoundCategory.PLAYERS, 1.0F, 1.0F / (itemRand.nextFloat() * 0.4F + 1.2F) + f * 0.5F);

                    if (!flag1 && !entityplayer.capabilities.isCreativeMode)
                    {
                        itemstack.shrink(1);

                        if (itemstack.isEmpty())
                        {
                            entityplayer.inventory.deleteStack(itemstack);
                        }
                    }

                    entityplayer.addStat(StatList.getObjectUseStats(this));
                }
            }
        }
    }

    /**
     * Gets the velocity of the arrow entity from the bow's charge
     */
    public static float getArrowVelocity(int charge)
    {
        float f = (float)charge / 20.0F;
        f = (f * f + f * 2.0F) / 3.0F;

        if (f > 1.0F)
        {
            f = 1.0F;
        }

        return f;
    }

    /**
     * How long it takes to use or consume an item
     */
    public int getMaxItemUseDuration(ItemStack stack)
    {
        return 72000;
    }

    /**
     * returns the action that specifies what animation to play when the items is being used
     */
    public EnumAction getItemUseAction(ItemStack stack)
    {
        return EnumAction.BOW;
    }

    /**
     * Called when the equipped item is right clicked.
     */
    public ActionResult<ItemStack> onItemRightClick(World worldIn, EntityPlayer playerIn, EnumHand handIn)
    {
        ItemStack itemstack = playerIn.getHeldItem(handIn);
        if (!(itemstack.getItem() instanceof ItemSpellBow))
            return new ActionResult<>(EnumActionResult.PASS, itemstack);

        NBTTagCompound nbt = UtilNbt.getTag(itemstack);
        boolean isBook = nbt.getBoolean(TAG_IS_IN_BOOK_MODE);
        if(playerIn.isSneaking()){
            nbt.setBoolean(TAG_IS_IN_BOOK_MODE,!isBook);
            return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
        }
        if(isBook)
            return doBookAction(worldIn, playerIn, itemstack);
        else
            return doBowAction(worldIn, playerIn, handIn, itemstack);
    }

    private ActionResult<ItemStack> doBookAction(World world, EntityPlayer player, ItemStack stack) {
        player.openGui(DivineFavor.instance, ConstGuiIDs.SPELL_BOW, world, (int) player.posX, (int) player.posY, (int) player.posZ);
        return new ActionResult<>(EnumActionResult.SUCCESS, stack);
    }

    private ActionResult<ItemStack> doBowAction(World worldIn, EntityPlayer playerIn, EnumHand handIn, ItemStack itemstack) {
        boolean gotArrows = !this.findAmmo(playerIn).isEmpty();

        ActionResult<ItemStack> ret = net.minecraftforge.event.ForgeEventFactory.onArrowNock(itemstack, worldIn, playerIn, handIn, gotArrows);
        if (ret != null) return ret;

        if (!playerIn.capabilities.isCreativeMode && !gotArrows)
        {
            return gotArrows ? new ActionResult(EnumActionResult.PASS, itemstack) : new ActionResult(EnumActionResult.FAIL, itemstack);
        }
        else
        {
            playerIn.setActiveHand(handIn);
            return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);
        }
    }

    /**
     * Return the enchantability factor of the item, most of the time is based on material.
     */
    public int getItemEnchantability()
    {
        return 1;
    }

    @Override
    public ICapabilityProvider initCapabilities(ItemStack item, NBTTagCompound nbt) {
        if (item.getItem() == ModItems.spell_bow)
            return new SpellBowProvider();
        return null;
    }

    @Override
    public boolean getShareTag() {
        return true;
    }

    @Nullable
    @Override
    public NBTTagCompound getNBTShareTag(ItemStack stack) {
        NBTTagCompound tag = super.getNBTShareTag(stack);
        if (tag == null)
            tag = new NBTTagCompound();

        ISpellBowHandler grimoireHandler = stack.getCapability(CAPABILITY_SPELL_BOW, null);
        assert grimoireHandler != null;
        NBTTagCompound tagShare = SpellBowStorage.getNbtBase(grimoireHandler);
        tag.setTag(TAG_SHARE, tagShare);
        return tag;
    }

    @Override
    public void readNBTShareTag(ItemStack stack, @Nullable NBTTagCompound nbt) {
        super.readNBTShareTag(stack, nbt);
        if (nbt == null)
            return;
        ISpellBowHandler grimoireHandler = stack.getCapability(CAPABILITY_SPELL_BOW, null);
        assert grimoireHandler != null;
        NBTTagCompound tagShare = nbt.getCompoundTag(TAG_SHARE);
        SpellBowStorage.readNbtBase(grimoireHandler, tagShare);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/item/talisman_tools/spell_bow/ItemSpellBow.kt
package aurocosh.divinefavor.common.item.talisman_tools.spell_bow

import aurocosh.divinefavor.DivineFavor
import aurocosh.divinefavor.common.config.common.ConfigItem
import aurocosh.divinefavor.common.constants.ConstGuiIDs
import aurocosh.divinefavor.common.constants.ConstMainTabOrder
import aurocosh.divinefavor.common.item.arrow_talismans.base.ItemArrowTalisman
import aurocosh.divinefavor.common.item.base.ModItem
import aurocosh.divinefavor.common.item.common.ModItems
import aurocosh.divinefavor.common.item.spell_talismans.context.CastContext
import aurocosh.divinefavor.common.item.spell_talismans.context.CastContextGenerator
import aurocosh.divinefavor.common.item.talisman.ISelectedStackProvider
import aurocosh.divinefavor.common.item.talisman.IStackContainerProvider
import aurocosh.divinefavor.common.item.talisman_tools.BookPropertyWrapper
import aurocosh.divinefavor.common.item.talisman_tools.IStackContainer
import aurocosh.divinefavor.common.item.talisman_tools.CastableAdapter
import aurocosh.divinefavor.common.item.talisman_tools.TalismanContainerMode
import aurocosh.divinefavor.common.item.talisman_tools.spell_bow.capability.SpellBowDataHandler.CAPABILITY_SPELL_BOW
import aurocosh.divinefavor.common.item.talisman_tools.spell_bow.capability.SpellBowProvider
import aurocosh.divinefavor.common.item.talisman_tools.spell_bow.capability.SpellBowStorage
import aurocosh.divinefavor.common.lib.extensions.cap
import aurocosh.divinefavor.common.lib.extensions.isNotEmpty
import aurocosh.divinefavor.common.stack_actions.StackAction
import aurocosh.divinefavor.common.stack_actions.StackActionHandler
import aurocosh.divinefavor.common.stack_actions.interfaces.IActionAccessor
import aurocosh.divinefavor.common.stack_actions.interfaces.IActionContainer
import aurocosh.divinefavor.common.stack_properties.StackPropertyHandler
import aurocosh.divinefavor.common.stack_properties.interfaces.IStackPropertyAccessor
import aurocosh.divinefavor.common.stack_properties.interfaces.IStackPropertyContainer
import aurocosh.divinefavor.common.util.UtilBow
import aurocosh.divinefavor.common.util.UtilItemStack.actionResult
import aurocosh.divinefavor.common.util.UtilPlayer
import net.minecraft.client.resources.I18n
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.enchantment.EnchantmentHelper
import net.minecraft.enchantment.EnumEnchantmentType
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.projectile.EntityArrow
import net.minecraft.entity.projectile.EntityTippedArrow
import net.minecraft.init.Enchantments
import net.minecraft.init.Items
import net.minecraft.init.SoundEvents
import net.minecraft.item.EnumAction
import net.minecraft.item.Item
import net.minecraft.item.ItemArrow
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.stats.StatList
import net.minecraft.util.*
import net.minecraft.world.World
import net.minecraftforge.common.capabilities.ICapabilityProvider
import net.minecraftforge.event.ForgeEventFactory
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import net.minecraftforge.oredict.OreDictionary

class ItemSpellBow : ModItem("spell_bow", "spell_bow/spell_bow", ConstMainTabOrder.CONTAINERS), ISelectedStackProvider, IStackContainerProvider, IStackPropertyContainer, IActionContainer {
    protected val propertyHandler: StackPropertyHandler = StackPropertyHandler("spell_bow")
    override val properties: IStackPropertyAccessor = propertyHandler
    protected val actionHandler: StackActionHandler = StackActionHandler("spell_bow")
    override val actions: IActionAccessor = actionHandler
    private val bookPropertyWrapper = BookPropertyWrapper(propertyHandler)
    private val material: ToolMaterial = ToolMaterial.WOOD

    override fun findProperty(stack: ItemStack, item: Item, propertyName: String) =
            CastableAdapter.findProperty(stack, item, propertyName, this, propertyHandler, CAPABILITY_SPELL_BOW)
    override fun findAction(stack: ItemStack, item: Item, actionName: String): Pair<ItemStack, StackAction>? =
            CastableAdapter.findAction(stack, item, actionName, this, actionHandler, CAPABILITY_SPELL_BOW)

    init {
        maxStackSize = 1
        creativeTab = DivineFavor.TAB_MAIN
        maxDamage = ConfigItem.spellBow.maxUses

        addPropertyOverride(ResourceLocation("book_mode")) { stack, _, _ -> bookPropertyWrapper.getValueForModel(stack) }
        addPropertyOverride(ResourceLocation("pulling")) { stack, _, entity -> if (entity != null && entity.isHandActive && entity.activeItemStack == stack) 1.0f else 0.0f }
        addPropertyOverride(ResourceLocation("pull")) { stack, _, entityLivingBase ->
            when {
                entityLivingBase == null -> 0.0f
                entityLivingBase.activeItemStack.item !== ModItems.spell_bow -> 0.0f
                else -> (stack.maxItemUseDuration - entityLivingBase.itemInUseCount).toFloat() / 20.0f
            }
        }
    }

    private fun findAmmo(player: EntityPlayer): ItemStack {
        return UtilPlayer.findStackInInventory(player) { stack -> stack.item is ItemArrow }.stack
    }

    /**
     * Called when the player stops using an Item (stops holding the right mouse button).
     */
    override fun onPlayerStoppedUsing(bowStack: ItemStack, world: World, entityLiving: EntityLivingBase, timeLeft: Int) {
        if (entityLiving !is EntityPlayer)
            return
        val mode = bookPropertyWrapper.getModeOrTransform(bowStack, entityLiving)
        if (mode != TalismanContainerMode.NORMAL)
            return

        val unlimitedArrows = entityLiving.capabilities.isCreativeMode || EnchantmentHelper.getEnchantmentLevel(Enchantments.INFINITY, bowStack) > 0
        var arrowStack = findAmmo(entityLiving)

        var charge = getMaxItemUseDuration(bowStack) - timeLeft
        charge = ForgeEventFactory.onArrowLoose(bowStack, world, entityLiving, charge, !arrowStack.isEmpty || unlimitedArrows)
        if (charge < 0)
            return
        if (arrowStack.isEmpty && !unlimitedArrows)
            return

        if (arrowStack.isEmpty)
            arrowStack = ItemStack(Items.ARROW)

        val velocity = UtilBow.getArrowVelocity(charge)
        if (velocity.toDouble() < 0.1)
            return

        val stackIsInfinite = entityLiving.capabilities.isCreativeMode || arrowStack.item is ItemArrow && (arrowStack.item as ItemArrow).isInfinite(arrowStack, bowStack, entityLiving)
        if (!world.isRemote) {

            val container = bowStack.item as ISelectedStackProvider
            val talismanStack = container.getSelectedStack(bowStack)

            val context = CastContextGenerator.arrowShot(world, entityLiving, talismanStack, bowStack)

            val (arrowTalisman, entityArrow) = getArrow(context, arrowStack)
            entityArrow.shoot(entityLiving, entityLiving.rotationPitch, entityLiving.rotationYaw, 0.0f, velocity * 3.0f, 1.0f)

            if (velocity == 1.0f)
                entityArrow.isCritical = true

            val powerLevel = EnchantmentHelper.getEnchantmentLevel(Enchantments.POWER, bowStack)
            if (powerLevel > 0)
                entityArrow.damage = entityArrow.damage + powerLevel.toDouble() * 0.5 + 0.5

            val punchLevel = EnchantmentHelper.getEnchantmentLevel(Enchantments.PUNCH, bowStack)
            if (punchLevel > 0)
                entityArrow.setKnockbackStrength(punchLevel)

            if (EnchantmentHelper.getEnchantmentLevel(Enchantments.FLAME, bowStack) > 0)
                entityArrow.setFire(100)

            bowStack.damageItem(1, entityLiving)

            if (stackIsInfinite || entityLiving.capabilities.isCreativeMode && (arrowStack.item === Items.SPECTRAL_ARROW || arrowStack.item === Items.TIPPED_ARROW))
                entityArrow.pickupStatus = EntityArrow.PickupStatus.CREATIVE_ONLY
            world.spawnEntity(entityArrow)
            arrowTalisman?.postInit(entityArrow)
        }

        world.playSound(null, entityLiving.posX, entityLiving.posY, entityLiving.posZ, SoundEvents.ENTITY_ARROW_SHOOT, SoundCategory.PLAYERS, 1.0f, 1.0f / (Item.itemRand.nextFloat() * 0.4f + 1.2f) + velocity * 0.5f)

        if (!stackIsInfinite && !entityLiving.capabilities.isCreativeMode)
            UtilPlayer.damageStack(entityLiving, arrowStack)
        entityLiving.addStat(StatList.getObjectUseStats(this)!!)
    }

    private fun getArrow(context: CastContext, arrowStack: ItemStack): Pair<ItemArrowTalisman?, EntityArrow> {
        val talisman = context.stack.item
        if (talisman !is ItemArrowTalisman || !talisman.claimCost(context))
            return Pair(null, getStandardArrow(context.world, arrowStack, context.player))
        val arrow = talisman.createArrow(context)
        return Pair(talisman, arrow)
    }

    private fun getStandardArrow(world: World, arrowStack: ItemStack, shooter: EntityLivingBase): EntityArrow {
        val arrow = EntityTippedArrow(world, shooter)
        arrow.setPotionEffect(arrowStack)
        return arrow
    }

    /**
     * How long it takes to use or consume an item
     */
    override fun getMaxItemUseDuration(stack: ItemStack?): Int {
        return 72000
    }

    /**
     * returns the action that specifies what animation to play when the items is being used
     */
    override fun getItemUseAction(stack: ItemStack?): EnumAction {
        return EnumAction.BOW
    }

    /**
     * Called when the equipped item is right clicked.
     */
    override fun onItemRightClick(world: World, player: EntityPlayer, hand: EnumHand): ActionResult<ItemStack> {
        val itemStack = player.getHeldItem(hand)
        if (itemStack.item !is ItemSpellBow)
            return ActionResult(EnumActionResult.PASS, itemStack)
        return when (bookPropertyWrapper.getModeOrTransform(itemStack, player)) {
            TalismanContainerMode.BOOK -> doBookAction(world, player, itemStack)
            TalismanContainerMode.NORMAL -> doBowAction(world, player, hand, itemStack)
            else -> return actionResult(false, itemStack)
        }
    }

    private fun doBookAction(world: World?, player: EntityPlayer, stack: ItemStack): ActionResult<ItemStack> {
        player.openGui(DivineFavor, ConstGuiIDs.SPELL_BOW, world!!, player.posX.toInt(), player.posY.toInt(), player.posZ.toInt())
        return ActionResult(EnumActionResult.SUCCESS, stack)
    }

    private fun doBowAction(worldIn: World?, playerIn: EntityPlayer, handIn: EnumHand, itemstack: ItemStack): ActionResult<ItemStack> {
        val gotArrows = !findAmmo(playerIn).isEmpty

        val ret = ForgeEventFactory.onArrowNock(itemstack, worldIn, playerIn, handIn, gotArrows)
        if (ret != null)
            return ret

        if (!playerIn.capabilities.isCreativeMode && !gotArrows)
            return if (gotArrows) ActionResult(EnumActionResult.PASS, itemstack) else ActionResult(EnumActionResult.FAIL, itemstack)
        else {
            playerIn.activeHand = handIn
            return ActionResult(EnumActionResult.SUCCESS, itemstack)
        }
    }

    /**
     * Return the enchantability factor of the item, most of the time is based on material.
     */
    override fun getItemEnchantability(): Int {
        return ConfigItem.spellBow.enchantability
    }

    override fun canApplyAtEnchantingTable(stack: ItemStack, enchantment: net.minecraft.enchantment.Enchantment): Boolean {
        return if(enchantment.type == EnumEnchantmentType.BOW)
            true;
        else
            super.canApplyAtEnchantingTable(stack, enchantment)
    }

    override fun initCapabilities(item: ItemStack, nbt: NBTTagCompound?): ICapabilityProvider? {
        return if (item.item === ModItems.spell_bow) SpellBowProvider() else null
    }

    override fun getIsRepairable(toRepair: ItemStack, repair: ItemStack): Boolean {
        val stack = material.repairItemStack
        if (stack.isEmpty)
            return super.getIsRepairable(toRepair, repair)
        if (!OreDictionary.itemMatches(stack, repair, false)) return super.getIsRepairable(toRepair, repair)
        return true
    }

    override fun getStackContainer(stack: ItemStack): IStackContainer = stack.cap(CAPABILITY_SPELL_BOW)

    override fun getSelectedStack(stack: ItemStack): ItemStack {
        if (stack.item !== this)
            return ItemStack.EMPTY
        return stack.cap(CAPABILITY_SPELL_BOW).getSelectedStack()
    }

    @SideOnly(Side.CLIENT)
    override fun addInformation(stack: ItemStack, world: World?, tooltip: MutableList<String>, flag: ITooltipFlag) {
        super.addInformation(stack, world, tooltip, flag)

        val talismanTool = getStackContainer(stack)
        val talismanCount = talismanTool.getAllStacks().filter (ItemStack::isNotEmpty).count()
        val countMessage = I18n.format("tooltip.divinefavor:talisman_tool.talisman_count", talismanCount)
        tooltip.add(countMessage)

        properties.getPropertyTooltip(stack).forEach { tooltip.add(it) }
    }

    override fun getShareTag(): Boolean {
        return true
    }

    override fun getNBTShareTag(stack: ItemStack): NBTTagCompound? {
        var tag = super.getNBTShareTag(stack)
        if (tag == null)
            tag = NBTTagCompound()

        val bowHandler = stack.cap(CAPABILITY_SPELL_BOW)
        val tagShare = SpellBowStorage.getNbtBase(bowHandler)
        tag.setTag(TAG_SHARE, tagShare)
        return tag
    }

    override fun readNBTShareTag(stack: ItemStack, nbt: NBTTagCompound?) {
        super.readNBTShareTag(stack, nbt)
        if (nbt == null)
            return
        val bowHandler = stack.cap(CAPABILITY_SPELL_BOW)
        val tagShare = nbt.getCompoundTag(TAG_SHARE)
        SpellBowStorage.readNbtBase(bowHandler, tagShare)
    }

    companion object {
        const val SLOT_COUNT = 27
        private const val TAG_SHARE = "SpellBowShare"
    }
}